src/rust/inetstack/protocols/layer4/tcp/established/receiver.rs

```diff
@@ -274,60 +186,166 @@ impl Receiver {
+        debug_assert!(
+            (self.window_scale_shift_bits == 0 && bytes_unread <= MAX_WINDOW_SIZE_WITHOUT_SCALING)
+                || bytes_unread <= MAX_WINDOW_SIZE_WITH_SCALING
+        );
+        debug!(
+            "Receive window size: bytes_unread={:?} buffer_size_bytes={:?} ",
+            bytes_unread, self.buffer_size_bytes
+        );
+        self.buffer_size_bytes - bytes_unread
+    }
+
+    pub fn hdr_window_size(&self) -> u16 {
+        let window_size: u32 = self.get_receive_window_size();
+        let hdr_window_size: u16 = expect_ok!(
+            (window_size >> self.window_scale_shift_bits).try_into(),
+            "Window size overflow"
+        );
+        debug!(
+            "Window size -> {} (hdr {}, scale {})",
+            (hdr_window_size as u32) << self.window_scale_shift_bits,
+            hdr_window_size,
+            self.window_scale_shift_bits,
+        );
+        hdr_window_size
+    }
+
+    // This routine takes an incoming in-order TCP segment and adds the data to the user's receive queue.  If the new
+    // segment fills a "hole" in the receive sequence number space allowing previously stored out-of-order data to now
+    // be received, it receives that too.
+    //
+    // This routine also updates receive_next to reflect any data now considered "received".
+    fn receive_data(&mut self, seg_start: SeqNumber, buf: DemiBuffer) {
+        // This routine should only be called with in-order segment data.
+        debug_assert_eq!(seg_start, self.receive_next_seq_no);
+
+        // Push the new segment data onto the end of the receive queue.
+        self.receive_next_seq_no = self.receive_next_seq_no + SeqNumber::from(buf.len() as u32);
+        // This inserts the segment and wakes a waiting pop coroutine.
+        self.pop_queue.push(buf);
+
+        // Okay, we've successfully received some new data.  Check if any of the formerly out-of-order data waiting in
+        // the out-of-order queue is now in-order.  If so, we can move it to the receive queue.
+        while !self.out_of_order_frames.is_empty() {
+            if let Some(stored_entry) = self.out_of_order_frames.front() {
+                if stored_entry.0 == self.receive_next_seq_no {
+                    // Move this entry's buffer from the out-of-order store to the receive queue.
+                    // This data is now considered to be "received" by TCP, and included in our RCV.NXT calculation.
+                    debug!("Recovering out-of-order packet at {}", self.receive_next_seq_no);
+                    if let Some(temp) = self.out_of_order_frames.pop_front() {
+                        self.receive_next_seq_no = self.receive_next_seq_no + SeqNumber::from(temp.1.len() as u32);
+                        // This inserts the segment and wakes a waiting pop coroutine.
+                        self.pop_queue.push(temp.1);
+                    }
+                } else {
```

Comment:
This conditional can be inverted to reduce nesting. The indentation level is pretty high in this function.

Reply:
Done! So much better!